CONTRIBUTING.md

```diff
@@ -1,136 +1,228 @@
+a pull request. This will enable a discussion about what is being implemented
+and how best we can work together to implement it in the code.
+
+If you've written some code but need help finishing it, want to get initial
+feedback on it priort to finishing it, or want to share it and discuss prior
+to finishing the implementation you can open a *Work in Progress* pull request.
+When you create the pull request prefix the title with the **\[WIP\]** tag (for
+work in progress). This will indicate to reviewers that the code in the PR
+isn't in it's final state and will change. It also means that we will not merge
+the commit until it is finished. You or a reviewer can remove the [WIP] tag when
+the code is ready to fully reviewed for merging.
 
 ### Contributor License Agreement
 
-We\'d love to accept your code! Before we can, we have to get a few
-legal requirements sorted out. By signing a contributor license
-agreement (CLA), we ensure that the community is free to use your
-contributions.
+Before you can submit any code we need all contributors to sign a
+contributor license agreement. By signing a contributor license
+agreement (CLA) you're basically just an attesting to the fact
+that you are the author of the contribution and that you're freely
+contributing it under the terms of the Apache-2.0 license.
 
 When you contribute to the Qiskit Terra project with a new pull request,
 a bot will evaluate whether you have signed the CLA. If required, the
 bot will comment on the pull request, including a link to accept the
 agreement. The [individual CLA] document is available for review as a PDF.
 
 **Note**:
-> If you work for a company that wants to allow you to contribute your
-> work, then you will need to sign a [corporate CLA] and email it to us
-> at <qiskit@us.ibm.com>.
+> If your contribution is part of your employment or your contribution
+> is the property of your employer, then you will likely need to sign a
+> [corporate CLA] too and email it to us at <qiskit@us.ibm.com>.
 
+#### Pull request checklist
 
-### Good first contributions
+When submitting a pull request and you feel it is ready for review,
+please ensure that:
 
-You are welcome to contribute wherever in the code you want to, of
-course, but we recommend taking a look at the [`good first
-contribution`](https://github.com/Qiskit/qiskit-terra/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) label into the issues and pick one. We would love to
-mentor you!
+1. The code follows the code style of the project and successfully
+   passes the tests. For convenience, you can execute `tox` locally,
+   which will run these checks and report any issues.
+2. The documentation has been updated accordingly. In particular, if a
+   function or class has been modified during the PR, please update the
+   *docstring* accordingly.
+3. If it makes sense for your change that you have added new tests that
```

Comment:
Should updating the CHANGELOG be included in this list?

Reply:
Done